csharp/ql/src/semmle/code/csharp/dataflow/FrameworkDataFlow.qll

```diff
@@ -0,0 +1,502 @@
+    }
+
+  newtype TSummaryInput =
+    TQualifierSummaryInput() or
+    TArgumentSummaryInput(int i) { i = any(Parameter p).getPosition() } or
+    TDelegateSummaryInput(int i) { hasDelegateArgumentPosition(_, i) }
+
+  newtype TSummaryOutput =
+    TQualifierSummaryOutput() or
+    TReturnSummaryOutput() or
+    TArgumentSummaryOutput(int i) {
+      exists(SourceDeclarationCallable c | exists(c.getParameter(i)))
+    } or
+    TDelegateSummaryOutput(int i, int j) { hasDelegateArgumentPosition2(_, i, j) }
+
+  newtype TFlowSource =
+    TRemoteFlowSource() or
+    TLocalFlowSource() or
+    TStoredFlowSource()
+
+  newtype TFlowSink =
+    TRemoteFlowSink() or
+    THtmlFlowSink() or
+    TEmailFlowSink() or
+    TSqlFlowSink() or
+    TCommandFlowSink()
+
+  /** Holds if a flow-source of kind `a` is also a flow-source of kind `b`. */
+  predicate flowSourceSubset(FlowSource a, FlowSource b) { none() }
+
+  /** Holds if a flow-sink of kind `a` is also a flow-sink of kind `b`. */
+  predicate flowSinkSubset(FlowSink a, FlowSink b) {
+    a in [THtmlFlowSink().(FlowSink), TEmailFlowSink().(FlowSink)] and
+    b = TRemoteFlowSink()
+  }
+}
+
+private import Internal
+
+/** A content stack. */
+class ContentStack extends TContentStack {
+  /** Gets the content stack obtained by popping `c` from this stack, if any. */
+  ContentStack pop(Content c) { this = TPushContentStack(c, result) }
+
+  /** Gets the length of this content stack. */
+  int length() {
+    this = TNilContentStack() and result = 0
+    or
+    result = 1 + this.pop(_).length()
+  }
+
+  /** Gets the content stack obtained by dropping the first `i` elements, if any. */
+  ContentStack drop(int i) {
+    i = 0 and result = this
+    or
+    result = this.pop(_).drop(i - 1)
+  }
+
+  /** Holds if this content stack contains content `c`. */
+  predicate contains(Content c) { exists(this.drop(_).pop(c)) }
+
+  /** Gets a textual representation of this content stack. */
+  string toString() {
+    exists(Content head, ContentStack tail |
+      tail = this.pop(head) and
+      if tail.length() = 0 then result = head.toString() else result = head + ", " + tail
+    )
+    or
+    this = TNilContentStack() and
+    result = "<empty>"
+  }
+}
+
+/** Provides predicates for constructing content stacks. */
+module ContentStack {
+  /** Gets the empty content stack. */
+  ContentStack empty() { result = TNilContentStack() }
+
+  /** Gets a singleton content stack containing `c`. */
+  ContentStack singleton(Content c) { result = TPushContentStack(c, TNilContentStack()) }
+
+  /** Gets the content stack obtained by pushing `head` onto `tail`. */
+  ContentStack push(Content head, ContentStack tail) { result = TPushContentStack(head, tail) }
+
+  /** Gets the singleton "element content" content stack. */
+  ContentStack element() { result = singleton(any(ElementContent c)) }
+
+  /** Gets a singleton property content stack. */
+  ContentStack property(Property p) {
```

Comment:
Since Property doesn't exist in other languages, I think this part should be moved to the language-private area, like the existing `DataFlow::Content` type

Reply:
Yes, the `ContentStack` module will contain language-specific entries. I think I will try to split up the file as we would in a real shared setting.